edgePy/DGEList.py

```diff
@@ -272,37 +307,144 @@ def library_size(self) -> np.array:
         """
         return np.sum(self.counts, 0)
 
-    def cpm(self, log: bool = False, prior_count: float = PRIOR_COUNT) -> None:
-        """Return the DGEList normalized to read counts per million."""
-        self.counts = 1e6 * self.counts / np.sum(self.counts, axis=0)
-        if log:
-            self.counts[self.counts == 0] = prior_count
-            self.counts = np.log(self.counts)
+    def log_transform(self, counts, prior_count):
+        """Compute the log of the counts"""
+        counts[counts == 0] = prior_count
+        return np.log(counts)
+
+    def cpm(self, transform_to_log: bool = False, prior_count: float = PRIOR_COUNT) -> "DGEList":
+        """Normalize the DGEList to read counts per million."""
+        counts = 1e6 * self.counts / np.sum(self.counts, axis=0)
+        current_log = self.current_log_status
+        if transform_to_log:
+            counts = self.log_transform(counts, prior_count)
+            current_log = True
+
+        return self.copy(counts=counts, current_log=current_log)
 
     def rpkm(
-        self, gene_lengths: Mapping, log: bool = False, prior_count: float = PRIOR_COUNT
-    ) -> None:
+        self,
+        gene_data: CanonicalDataStore,
+        transform_to_log: bool = False,
+        prior_count: float = PRIOR_COUNT,
+    ) -> "DGEList":
         """Return the DGEList normalized to reads per kilobase of gene length
-        per million reads.
+        per million reads. (RPKM =   numReads / ( geneLength/1000 * totalNumReads/1,000,000 )
 
+        Args:
+            gene_data: An object that works to import Ensembl based data, for use in calculations
+            transform_to_log: true, if you wish to convert to log after converting to RPKM
+            prior_count: a minimum value for genes, if you do log transforms.
         """
-        raise NotImplementedError
+        current_log = self.current_log_status
+
+        if self.current_log_status:
+            self.counts = np.exp(self.counts)
+            current_log = False
+        col_sum = np.sum(self.counts, axis=0)
+
+        gene_len_ordered, gene_mask = self.get_gene_mask_and_lengths(gene_data)
```

Comment:
CanonicalDataStore is already a class. We could extend it to provide functions like canonical_transcript_length (instead of defining them here). This way users had a way to access such features independently of (rpkm) normalization.

Reply:
What is gene_mask for btw?

Reply:
Gene mask is the list of genes you should keep, in the form of a mask which can be applied efficiently to the numpy dataset.  (It is an array of True/False, telling you which genes to keep)

Also very happy to keep regorganizing and adding to the CanonicalDataStore.  (That as the idea, really, but baby steps!)

Reply:
True! That can be done in a later PR also.